src/stp/plays/defensive/DefendShot.cpp

```diff
@@ -22,53 +22,111 @@ DefendShot::DefendShot() : Play() {
+    flagMap.insert({"midfielder_3", {DealerFlagPriority::MEDIUM_PRIORITY, {notImportant}}});
+    flagMap.insert({"midfielder_4", {DealerFlagPriority::MEDIUM_PRIORITY, {notImportant}}});
 
     return flagMap;
 }
 
 void DefendShot::calculateInfoForRoles() noexcept {
+    calculateInfoForWallers();
     calculateInfoForDefenders();
     calculateInfoForHarassers();
     calculateInfoForKeeper();
 }
 
+void DefendShot::calculateInfoForWallers() noexcept {
+    stpInfos["waller_1"].setPositionToDefend(field.getOurTopGoalSide());
+    stpInfos["waller_1"].setBlockDistance(BlockDistance::CLOSE);
+
+    stpInfos["waller_2"].setPositionToDefend(field.getOurGoalCenter());
+    stpInfos["waller_2"].setBlockDistance(BlockDistance::CLOSE);
+
+    stpInfos["waller_3"].setPositionToDefend(field.getOurBottomGoalSide());
+    stpInfos["waller_3"].setBlockDistance(BlockDistance::CLOSE);
+}
+
 void DefendShot::calculateInfoForDefenders() noexcept {
-    stpInfos["defender_1"].setPositionToDefend(field.getOurTopGoalSide());
+    auto enemyRobots = world->getWorld()->getThem();
+
+    auto enemyClosestToBall = world->getWorld()->getRobotClosestToBall(world::them);
+
+    erase_if(enemyRobots, [&](const auto enemyRobot) -> bool { return enemyClosestToBall && enemyRobot->getId() == enemyClosestToBall.value()->getId(); });
+
+    auto enemyClosestToOurGoalOne = world->getWorld()->getRobotClosestToPoint(field.getOurGoalCenter(), enemyRobots);
+
+    erase_if(enemyRobots, [&](const auto enemyRobot) -> bool { return enemyClosestToOurGoalOne && enemyRobot->getId() == enemyClosestToOurGoalOne.value()->getId(); });
+
+    auto enemyClosestToOurGoalTwo = world->getWorld()->getRobotClosestToPoint(field.getOurGoalCenter(), enemyRobots);
+
+    erase_if(enemyRobots, [&](const auto enemyRobot) -> bool { return enemyClosestToOurGoalTwo && enemyRobot->getId() == enemyClosestToOurGoalTwo.value()->getId(); });
+
+    stpInfos["defender_1"].setPositionToDefend(field.getOurGoalCenter());
+    stpInfos["defender_1"].setEnemyRobot(enemyClosestToOurGoalOne);
     stpInfos["defender_1"].setBlockDistance(BlockDistance::CLOSE);
 
-    stpInfos["defender_2"].setPositionToDefend(field.getOurBottomGoalSide());
+    stpInfos["defender_2"].setPositionToDefend(field.getOurGoalCenter());
+    stpInfos["defender_2"].setEnemyRobot(enemyClosestToOurGoalTwo);
     stpInfos["defender_2"].setBlockDistance(BlockDistance::CLOSE);
+
+    std::map<double, Vector2> enemyMap;
+
+    for (auto enemy : enemyRobots) {
+        double score = FieldComputations::getDistanceToGoal(field, true, enemy->getPos());
+        enemyMap.insert({score, enemy->getPos()});
+    }
+
+    for (int i = 1; i <= 4; i++) {
+        stpInfos["midfielder_" + std::to_string(i)].setPositionToDefend(enemyMap.begin()->second);
+        stpInfos["midfielder_" + std::to_string(i)].setBlockDistance(BlockDistance::CLOSE);
+
+        if (!enemyMap.empty()) {
```

Comment:
Same as in defendPass, this check is too late since line 115 still uses the possibly empty map